controllers/keystoneapi_controller.go

```diff
@@ -99,8 +100,7 @@ type KeystoneAPIReconciler struct {
 
 // Reconcile reconcile keystone API requests
 func (r *KeystoneAPIReconciler) Reconcile(ctx context.Context, req ctrl.Request) (result ctrl.Result, _err error) {
-	_ = r.Log.WithValues("keystoneapi", req.NamespacedName)
-
+	r.Log = log.FromContext(ctx)
```

Comment:
We should not do that as per:

> One thing come up while discussed this work with @SeanMooney. The actual Reconciler.Reconcile() function can be called in parallel if [MaxConcurrentReconciles](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/controller#Options) is set to > 1. This means that the Reconciler.Log might be written by two parallel Reconcile request. As we store the reconcile ID in the Logger (coming from the ctx) this could lead to mixing logs between parallel Reconcile requests. :/ This is a bummer as we need to pass along something to each function to access the logger as we cannot use the Reconciler receiver as a common storage. As we already have ctx in most of the calls I suggest to use that and the logger.FromContext(ctx) to access the logger from the ctx.

Reply:
So whenever the code logs (or passes the logger around, e.g. to the Helper) we need to use the logger stored in the cxt. You can even remove the Log field from the Reconcile struct to force all the places currently using r.Log to change.

Reply:
so something like this: ?

```go
diff --git a/controllers/keystoneapi_controller.go b/controllers/keystoneapi_controller.go
 type KeystoneAPIReconciler struct {
        client.Client
        Kclient kubernetes.Interface
-       Log     logr.Logger
        Scheme  *runtime.Scheme
 }

 // Reconcile reconcile keystone API requests
 func (r *KeystoneAPIReconciler) Reconcile(ctx context.Context, req ctrl.Request) (result ctrl.Result, _err error) {
-       r.Log = log.FromContext(ctx)
+       Log := log.FromContext(ctx)

 func (r *KeystoneAPIReconciler) reconcileDelete(ctx context.Context, instance *keystonev1.KeystoneAPI, helper *helper.Helper) (ctrl.Result, error) {
-       r.Log.Info("Reconciling Service delete")
-
+       helper.GetLogger().Info("Reconciling Service delete")
```

Reply:
The first two changes are OK to me. For the last one I would not relay on the helper. (Personally I don't like our Helper as it is really a catch all and does too much). Instead I would do 

```diff 
 func (r *KeystoneAPIReconciler) reconcileDelete(ctx context.Context, instance *keystonev1.KeystoneAPI, helper *helper.Helper) (ctrl.Result, error) {
-       r.Log.Info("Reconciling Service delete")
-
+       log.FromContext(ctx).Info("Reconciling Service delete")
```

This is not more complex than the the helper based solution but actually avoids the helper so eventually we can shrink (or even dissolve) the helper.

If a function logs a lot then you can obviously do `l := log.FromContext(ctx)` at the beginning of the function and use `l.Info()` later.